src/detect-ipaddr.c

```diff
@@ -0,0 +1,154 @@
+#endif
+
+    sigmatch_table[DETECT_IPADDR_SRC].flags |= SIGMATCH_NOOPT | SIGMATCH_INFO_STICKY_BUFFER;
+
+    g_src_ipaddr_buffer_id = DetectBufferTypeRegister(KEYWORD_NAME_SRC);
+    BUG_ON(g_src_ipaddr_buffer_id < 0);
+
+    DetectBufferTypeSupportsPacket(KEYWORD_NAME_SRC);
+
+    DetectPktMpmRegister(KEYWORD_NAME_SRC, 2, PrefilterGenericMpmPktRegister, GetDataSrc);
+
+    DetectPktInspectEngineRegister(
+            KEYWORD_NAME_SRC, GetDataSrc, DetectEngineInspectPktBufferGeneric);
+
+    sigmatch_table[DETECT_IPADDR_DST].name = KEYWORD_NAME_DST;
+    sigmatch_table[DETECT_IPADDR_DST].desc = "Sticky buffer for dest_ip";
+    sigmatch_table[DETECT_IPADDR_DST].url = "/rules/ipaddr.html#ip-dst";
+    sigmatch_table[DETECT_IPADDR_DST].Setup = DetectDestIPAddrBufferSetup;
+
+    sigmatch_table[DETECT_IPADDR_DST].flags |= SIGMATCH_NOOPT | SIGMATCH_INFO_STICKY_BUFFER;
+
+    g_dest_ipaddr_buffer_id = DetectBufferTypeRegister(KEYWORD_NAME_DST);
+    BUG_ON(g_dest_ipaddr_buffer_id < 0);
+
+    DetectBufferTypeSupportsPacket(KEYWORD_NAME_DST);
+
+    DetectPktMpmRegister(KEYWORD_NAME_DST, 2, PrefilterGenericMpmPktRegister, GetDataDst);
+
+    DetectPktInspectEngineRegister(
+            KEYWORD_NAME_DST, GetDataDst, DetectEngineInspectPktBufferGeneric);
+
+    /* NOTE: You may want to change this to SCLogNotice during development. */
+    SCLogDebug("IPAddr detect registered.");
+}
+
+static int DetectSrcIPAddrBufferSetup(DetectEngineCtx *de_ctx, Signature *s, const char *str)
+{
+    /* store list id. Content, pcre, etc will be added to the list at this
+     * id. */
+    s->init_data->list = g_src_ipaddr_buffer_id;
+
+    return 0;
+}
+
+static int DetectDestIPAddrBufferSetup(DetectEngineCtx *de_ctx, Signature *s, const char *str)
+{
+    /* store list id. Content, pcre, etc will be added to the list at this
+     * id. */
+    s->init_data->list = g_dest_ipaddr_buffer_id;
+
+    return 0;
+}
+
+/** \internal
+ *  \brief get the data to inspect from the buffer
+ *
+ *  \retval buffer or NULL in case of error
+ */
+static InspectionBuffer *GetDataSrc(DetectEngineThreadCtx *det_ctx,
+        const DetectEngineTransforms *transforms, Packet *p, const int list_id)
+{
+    InspectionBuffer *buffer = InspectionBufferGet(det_ctx, list_id);
+    if (buffer->inspect == NULL) {
+        InspectionBufferSetup(det_ctx, list_id, buffer, p->src.address.address_un_data8, 16);
+        InspectionBufferApplyTransforms(buffer, transforms);
+    }
+
+    return buffer;
+}
+
+/** \internal
+ *  \brief get the data to inspect from the buffer
+ *
+ *  \retval buffer or NULL in case of error
+ */
+static InspectionBuffer *GetDataDst(DetectEngineThreadCtx *det_ctx,
+        const DetectEngineTransforms *transforms, Packet *p, const int list_id)
+{
+    InspectionBuffer *buffer = InspectionBufferGet(det_ctx, list_id);
+    if (buffer->inspect == NULL) {
+        InspectionBufferSetup(det_ctx, list_id, buffer, p->dst.address.address_un_data8, 16);
```

Comment:
so the buffer size (bsize keyword) will be 16 even for ipv4?

Reply:
also matching on this for ipv4, would this work?
`ip.src; content:"|00 00 00 00 00 00|";`

Reply:
yes, ipv4 storage in ip set takes as much as IPv6 one.

Reply:
it is a binary field so the content match should work.

Reply:
I don't agree with this. Having 16 bytes padded with 12 0 bytes in case of IPv4 makes no sense.

Reply:
Does not make sense for matching or for memory usage ?

Reply:
For matching. We shouldn't expose nonsensical data to the rule lang. I think the simple solution is just to expose 4 bytes for IPv4, 16 for IPv6.

Reply:
I found a way to do that. It will be in the next MR.